src/backend/distributed/commands/foreign_constraint.c

```diff
@@ -561,106 +516,72 @@ get_relation_constraint_oid_compat(HeapTuple heapTuple)
 
 
 /*
- * HasForeignKeyToReferenceTable function scans the pgConstraint table to
- * fetch all of the constraints on the given relationId and see if at least one
- * of them is a foreign key referencing to a reference table.
+ * HasForeignKeyToReferenceTable function returns true if any of the foreign
+ * key constraints on the relation with relationOid references to a reference
+ * table.
  */
 bool
-HasForeignKeyToReferenceTable(Oid relationId)
+HasForeignKeyToReferenceTable(Oid relationOid)
 {
-	ScanKeyData scanKey[1];
-	int scanKeyCount = 1;
-	bool hasForeignKeyToReferenceTable = false;
+	List *tableForeignKeyOidsToReferenceTables =
+		GetForeignKeyOidsToReferenceTables(relationOid);
 
-	Relation pgConstraint = heap_open(ConstraintRelationId, AccessShareLock);
-	ScanKeyInit(&scanKey[0], Anum_pg_constraint_conrelid, BTEqualStrategyNumber, F_OIDEQ,
-				relationId);
-	SysScanDesc scanDescriptor = systable_beginscan(pgConstraint,
-													ConstraintRelidTypidNameIndexId,
-													true, NULL,
-													scanKeyCount, scanKey);
+	return list_length(tableForeignKeyOidsToReferenceTables) > 0;
+}
 
-	HeapTuple heapTuple = systable_getnext(scanDescriptor);
-	while (HeapTupleIsValid(heapTuple))
+
+/*
+ * GetForeignKeyOidsToReferenceTables function returns list of OIDs for the
+ * foreign key constraints on the given relationOid that are referencing to
+ * reference tables.
+ */
+static List *
+GetForeignKeyOidsToReferenceTables(Oid relationOid)
+{
+	int flags = EXTRACT_REFERENCING_CONSTRAINTS;
+	List *foreignKeyOids = GetTableForeignConstraintOidsInternal(relationOid, flags);
+
+	List *fkeyOidsToReferenceTables = NIL;
+
+	Oid foreignKeyOid = InvalidOid;
+	foreach_oid(foreignKeyOid, foreignKeyOids)
 	{
-		Form_pg_constraint constraintForm = (Form_pg_constraint) GETSTRUCT(heapTuple);
+		HeapTuple heapTuple = SearchSysCache1(CONSTROID, ObjectIdGetDatum(
+												  foreignKeyOid));
 
-		if (constraintForm->contype != CONSTRAINT_FOREIGN)
-		{
-			heapTuple = systable_getnext(scanDescriptor);
-			continue;
-		}
+		Assert(HeapTupleIsValid(heapTuple));
 
-		Oid referencedTableId = constraintForm->confrelid;
+		Form_pg_constraint constraintForm = (Form_pg_constraint) GETSTRUCT(heapTuple);
+
+		Oid referencedTableOid = constraintForm->confrelid;
 
-		if (!IsCitusTable(referencedTableId))
+		if (!IsReferenceTable(referencedTableOid))
```

Comment:
minor: maybe avoid calling `ReleaseSysCache` in multiple places:
```C
		Oid referencedTableOid = constraintForm->confrelid;

		if (IsReferenceTable(referencedTableOid))
		{
			fkeyOidsToReferenceTables = lappend_oid(fkeyOidsToReferenceTables, foreignKeyOid);
		}

		ReleaseSysCache(heapTuple);
```

(it sometimes become pain to maintain these )